src/device/intrinsics/random.jl

```diff
@@ -0,0 +1,49 @@
+function get_global_thread_index()
+    block_id = (gridDim().x * (blockIdx().y - 1)) + (blockIdx().x - 1)
+
+    thread_id = (block_id * (blockDim().x * blockDim().y)) +
+                ((threadIdx().y - 1) * blockDim().x) +
+                (threadIdx().x - 1)
+
+    return thread_id + 1
+end
+
+function xorshift(x::UInt32)::UInt32
+    x = xor(x, x << 13)
+    x = xor(x, x >> 17)
+    x = xor(x, x << 5)
+    return x
+end
+
+@eval @inline global_random_state() =
+    Base.llvmcall(
+        $("""@global_random_state = weak externally_initialized global i$(WORD_SIZE) 0
+             define i64 @entry() #0 {
+                 %ptr = load i$(WORD_SIZE), i$(WORD_SIZE)* @global_random_state, align 8
+                 ret i$(WORD_SIZE) %ptr
+             }
+             attributes #0 = { alwaysinline }
+          """, "entry"), Ptr{Cvoid}, Tuple{})
+
+function rand()
+    global_state_address = global_random_state()
+    global_state_ptr = LLVMPtr{UInt32, AS.Generic}(global_state_address)
+    index = get_global_thread_index()
+
+    state = unsafe_load(global_state_ptr, index, Val(8))
```

Comment:
Better use `datatype_alignment` here. Maybe it's better to just create a `CuDeviceVector{ThreadLocalRNG}` here, to reuse that functionality and not require these unsafe ops.